django_refugeerights/snappyuploader/tasks.py

```diff
@@ -1,41 +1,194 @@
+        py_data=data)
+    # Set topic_id to new topic id
+    return response
+
+
+def is_valid_snappy_data(csv_entry):
+    for field in ["lang", "topic", "question", "answer"]:
+        if not field in csv_entry or csv_entry[field] == "":
+            return False
+    return True
+
+
+def record_failure(csv_entry):
+    fl = FailureLog()
+    fl.csv_content = csv_entry
+    fl.save()
+    return "Failure recorded in DB."
+
+
+class Post_FAQ(Task):
+
+    """
+    Task to add an FAQ entry to Snappy via Api post
+    """
+    name = "snappyuploader.tasks.post_faq"
+
+    def run(self, topic_id, faq_id, data, **kwargs):
+        """
+        Returns FAQ id of imported FAQ
+        """
+        response = snappy_request(
+            "POST",
+            "account/%s/faqs/%s/topics/%s/questions" % (
+                settings.SNAPPY_ACCOUNT_ID, faq_id, topic_id),
+            py_data=data)
+
+        return "Uploaded FAQ %s" % response["id"]
+
+post_faq = Post_FAQ()
+
+
+class CSV_Importer(Task):
+
+    """
+    Task to take dict import the data
+    """
+    name = "snappyuploader.tasks.csv_importer"
+
+    class FailedEventRequest(Exception):
+
+        """
+        The attempted task failed because of a non-200 HTTP return
+        code.
+        """
+
+    def run(self, csv_data, faq_id, **kwargs):
+        """
+        Returns count of imported faqs
+        """
+        l = self.get_logger(**kwargs)
+
+        l.info("Processing new snappy FAQ import data")
+        faqs_added = 0
+        topics_added = 0
+        failed_imports = 0
+        try:
+            # Get FAQ topics
+            snappy_topics_data = snappy_request(
+                "GET",
+                "account/%s/faqs/%s/topics" % (
+                    settings.SNAPPY_ACCOUNT_ID, faq_id, ))
+            # Save the topics in a reusable dict {topic1: id1, topic2: id2,}
+            topics_map = {}
+            for snappy_topic_obj in snappy_topics_data:
+                topics_map[snappy_topic_obj["topic"]] = snappy_topic_obj["id"]
+
+            # Upload the CSV entries line by line
+            for csv_entry in csv_data:
+                if is_valid_snappy_data(csv_entry):
+                    topic_title = "[%s]%s" % (csv_entry["lang"],
+                                              csv_entry["topic"])
+                    if topic_title in topics_map:
+                        topic_id = topics_map[topic_title]
+                    else:
+                        # Post new topic to snappy
+                        result = add_snappy_topic(topic_title, faq_id)
+                        topics_added += 1
+                        topic_id = result["id"]
+                        # Add new topic to topics_map
+                        topics_map[topic_title] = topic_id
+
+                    # Post new FAQ
+                    data = {"question": (csv_entry["lang"] + csv_entry["question"]),
+                            "answer": csv_entry["answer"]}
+                    try:
```

Comment:
This try isnt' going to do anything because the post_faq is a async task -- move that inside the task